src/main/java/htsjdk/samtools/reference/FastaSequenceIndexCreator.java

```diff
@@ -0,0 +1,179 @@
+ *
+ * @author Daniel Gomez-Sanchez (magicDGS)
+ */
+public final class FastaSequenceIndexCreator {
+
+    // cannot be instantiated because it is an utility class
+    private FastaSequenceIndexCreator() {}
+
+    /**
+     * Creates a FASTA .fai index for the provided FASTA.
+     *
+     * @param fastaFile the file to build the index from.
+     *
+     * @throws SAMException if the fai file already exists or the file is malformed.
+     * @throws IOException  if an IO error occurs.
+     */
+    public static void create(final Path fastaFile) throws IOException {
+        // get the index to write the file in
+        final Path indexFile = ReferenceSequenceFileFactory.getFastaIndexFileName(fastaFile);
+        if (Files.exists(indexFile)) {
+            // throw an exception if the file already exists
+            throw new SAMException("Index file " + indexFile + " already exists for " + fastaFile);
+        }
+        // build the index
+        final FastaSequenceIndex index = buildFromFasta(fastaFile);
+        index.write(indexFile);
+    }
+
+    /**
+     * Builds a FastaSequenceIndex on the fly from a FASTA file.
+     *
+     * Note: this also alows to create an index for a compressed file, but does not generate the
+     * .gzi index required for use it with samtools.
+     *
+     * @param fastaFile the FASTA file.
+     *
+     * @return a fai index.
+     *
+     * @throws SAMException for formatting errors.
+     * @throws IOException  if an IO error occurs.
+     */
+    public static FastaSequenceIndex buildFromFasta(final Path fastaFile) throws IOException {
+        try(final AsciiLineReader in = new AsciiLineReader(IOUtil.openFileForReading(fastaFile))) {
+
+            // sanity check reference format:
+            // 1. Non-empty file
+            // 2. Header name starts with >
+            String previous = in.readLine();
+            if (previous == null) {
+                throw new SAMException("Cannot index empty file: " + fastaFile);
+            } else if (previous.charAt(0) != '>') {
+                throw new SAMException("Wrong sequence header: " + previous);
+            }
+
+            // initialize the sequence index
+            int sequenceIndex = -1;
+            // the location should be kept before iterating over the rest of the lines
+            long location = in.getPosition();
+
+            // initialize an empty index and the entry builder to null
+            final FastaSequenceIndex index = new FastaSequenceIndex();
+            FaiEntryBuilder entry = null;
+
+            // read the lines two by two
+            for (String line = in.readLine(); previous != null; line = in.readLine()) {
+                // in this case, the previous line contains a header and the current line the first sequence
+                if (previous.charAt(0) == '>') {
+                    // first entry should be skipped
+                    if (entry != null) {
```

Comment:
This is my personal taste, but I always prefer brackets. Anyway I changed it, because the style for htsjdk allows it.